Source/ASDisplayNode.mm

```diff
@@ -966,22 +959,26 @@ - (ASLayout *)calculateLayoutThatFits:(ASSizeRange)constrainedSize
   ASDN::MutexLocker l(__instanceLock__);
 
 #if YOGA_TREE_CONTIGUOUS /* YOGA */
-  if (ASHierarchyStateIncludesYogaLayoutEnabled(_hierarchyState) == YES) {
-    if (ASHierarchyStateIncludesYogaLayoutMeasuring(_hierarchyState) == NO && self.yogaCalculatedLayout == nil) {
-      ASDN::MutexUnlocker ul(__instanceLock__);
+  // There are several cases where Yoga could arrive here:
+  // - This node is not in a Yoga tree: it has neither a yogaParent nor yogaChildren.
+  // - This node is a Yoga tree root: it has no yogaParent, but has yogaChildren.
+  // - This node is a Yoga tree node: it has both a yogaParent and yogaChildren.
+  // - This node is a Yoga tree leaf: it has a yogaParent, but no yogaChidlren.
+  // If we're a leaf node, we are probably being called by a measure function and proceed as normal.
+  // If we're a yoga root or tree node, initiate a new Yoga calculation pass from root.
+  YGNodeRef yogaNode = _style.yogaNode;
+  if (yogaNode && (_yogaParent == nil || _yogaChildren.count > 0)) {
+    // This node has some connection to a Yoga tree.
+    ASDN::MutexUnlocker ul(__instanceLock__);
+    if (self.yogaLayoutInProgress == NO) {
       [self calculateLayoutFromYogaRoot:constrainedSize];
     }
-
-    // The call above may set yogaCalculatedLayout, even if it tested as nil to enter it.
-    if (self.yogaCalculatedLayout && self.yogaChildren.count > 0) {
-      return self.yogaCalculatedLayout;
-    }
+    return _yogaCalculatedLayout;
   }
 #endif /* YOGA */
   
   // Manual size calculation via calculateSizeThatFits:
-  if (((_methodOverrides & ASDisplayNodeMethodOverrideLayoutSpecThatFits) ||
-      (_layoutSpecBlock != NULL)) == NO) {
+  if (_layoutSpecBlock == NULL && (_methodOverrides & ASDisplayNodeMethodOverrideLayoutSpecThatFits) == 0) {
```

Comment:
This condition is equivalent to the previous one, but easier to read the intention of.